File: 1010/Main.java
import java.util.Scanner;

public class Main {
	public static void main(String[] args) {
		// 任意进制转十进制
		// Integer.parseInt(String number, radix)
		// System.out.println(Integer.parseInt("AB", 16));
		// BigInteger 处理大数
		// new BigInteger(String number. radix)
		// System.out.println(toDecimal("ab", 16));
		Scanner scanner = new Scanner(System.in);
		String astr = scanner.next();
		String bstr = scanner.next();
		int tag = scanner.nextInt();
		long radix = scanner.nextInt();
		long a;
		if (tag == 2) {
			String c = astr;
			astr = bstr;
			bstr = c;
		}
		a = toDecimal(astr, radix);
		long low = getMinRadix(bstr);
		long high = Math.max(a + 1, low);
		while (low <= high) {
			long mid = (low + high) / 2;
			if (a == toDecimal(bstr, mid))
			{
				System.out.print(mid);
				return ;
			} else if (a < toDecimal(bstr, mid)){
				high = mid - 1;
			} else {
				low = mid + 1;
			}
		}
		System.out.print("Impossible");
	}
	
	public static long toDecimal(String num, long radix) {
		long a = 0;
		for (int i = 0; i < num.length(); i++) {
			char c = num.charAt(num.length() - 1 - i);
			if (c >= '0' && c <= '9') {
				a += (c - '0') * Math.pow(radix, i);
			} else {
				a += (c - 'a' + 10) * Math.pow(radix, i);
			}
		}
		return a;
	}
	
	// 参考 https://blog.csdn.net/baidu_25107219/article/details/46700669
	// 需要考虑特殊情况，如 12 c 1 10
	// 进制至少要比出现的最大 digit 大 1
	public static long getMinRadix(String num) {
		int minRadix = 1;
		for (int i = 0; i < num.length(); i++) {
			char c = num.charAt(i);
			if (c >= '0' && c <= '9') {
				int r = c - '0';
				if (minRadix < r)
					minRadix = r;
			} else {
				int r = c - 'a' + 10;
				if (minRadix < r)
					minRadix = r;
			}
		}
		return minRadix + 1;
	}
}

File: 1010/main.cpp
#include <cstdio>
#include <cstring>
#include <algorithm>

using namespace std;

char N1[16];
char N2[16];

int tag;
unsigned long long radix;

unsigned long long to_deci(char number[16], unsigned long long radix)
{
    unsigned long long d = 0;
    unsigned long long power = 1;
    for (unsigned long long i = 0; i < strlen(number); i++)
    {
        unsigned long long k = number[i] >= 'a' ? number[i] - 'a' + 10 : number[i] - '0';
        d += k * power;
        power *= radix;
    }
    return d;
}

void revstr(char c[16])
{
    for (unsigned long long i = 0; i < strlen(c) / 2; i++)
    {
        unsigned long long ch = c[i];
        c[i] = c[strlen(c) - 1 - i];
        c[strlen(c) - 1 - i] = ch;
    }
}

unsigned long long get_minradix(char c[16])
{
    unsigned long long min_r = c[0] >= 'a' ? c[0] - 'a' + 10 : c[0] - '0';
    for (unsigned long long i = 1; i < strlen(c); i++)
    {
        unsigned long long t = c[i] >= 'a' ? c[i] - 'a' + 10 : c[i] - '0';
        min_r = t < min_r ? t : min_r;
    }
    return min_r + 1;
}

int main()
{
    scanf("%s %s %d %lld", N1, N2, &tag, &radix);
    revstr(N1);
    revstr(N2);
    if (tag != 1)
    {
        char tmpchar[16];
        strcpy(tmpchar, N1);
        strcpy(N1, N2);
        strcpy(N2, tmpchar);
    }

    unsigned long long k = to_deci(N1, radix);

    unsigned long long low = get_minradix(N2);
    unsigned long long high = max(k + 1, get_minradix(N2));
    unsigned long long mid;

    while (low <= high)
    {
        mid = (low + high) / 2;
        unsigned long long k0 = to_deci(N2, mid);
        if (k == k0)
        {
            printf("%llu", mid);
            return 0;
        } else if (k < k0)
            high = mid - 1;
        else
            low = mid + 1;
    }

    /*for (long long i = get_minradix(N2); i <= max(k + 1, get_minradix(N2)); i++)
    {
        long long k0 = to_deci(N2, i);
        if (k == k0)
        {
            printf("%lld", i);
            return 0;
        }
    }*/

    printf("Impossible");

    return 0;
}
